src/target/ext/tracepoints.rs

```diff
@@ -0,0 +1,523 @@
+    /// Note that `mask` may be larger than the word length.
+    Registers { mask: ManagedSlice<'a, u8> },
+    /// Collect `len` bytes of memory starting at the address in register number
+    /// `basereg`, plus `offset`. If `basereg` is None, then treat it as a fixed
+    /// address.
+    Memory {
+        basereg: Option<u64>,
+        offset: U,
+        length: u64,
+    },
+    /// Evaluate `expr`, which is a GDB agent bytecode expression, and collect
+    /// memory as it directs.
+    Expression { expr: ManagedSlice<'a, u8> },
+}
+
+#[cfg(feature = "alloc")]
+impl<'a, U: Copy> TracepointAction<'a, U> {
+    /// Allocate an owned copy of this structure.
+    pub fn get_owned<'b>(&self) -> TracepointAction<'b, U> {
+        use core::ops::Deref;
+        match self {
+            TracepointAction::Registers { mask } => TracepointAction::Registers {
+                mask: ManagedSlice::Owned(mask.deref().into()),
+            },
+            TracepointAction::Memory {
+                basereg,
+                offset,
+                length,
+            } => TracepointAction::Memory {
+                basereg: *basereg,
+                offset: *offset,
+                length: *length,
+            },
+            TracepointAction::Expression { expr } => TracepointAction::Expression {
+                expr: ManagedSlice::Owned(expr.deref().into()),
+            },
+        }
+    }
+}
+
+impl<'a, U: crate::internal::BeBytes + num_traits::Zero + PrimInt> TracepointAction<'a, U> {
+    pub(crate) fn write<C: Connection>(
+        &self,
+        res: &mut ResponseWriter<'_, C>,
+    ) -> Result<(), ResponseWriterError<C::Error>> {
+        match self {
+            TracepointAction::Registers { mask } => {
+                res.write_str("R ")?;
+                res.write_hex_buf(mask)?;
+            }
+            TracepointAction::Memory {
+                basereg,
+                offset,
+                length,
+            } => {
+                res.write_str("M ")?;
+                match basereg {
+                    Some(r) => res.write_num(*r),
+                    None => res.write_str("-1"),
+                }?;
+                res.write_str(",")?;
+                let mut buf = [0; 8];
+                offset.to_be_bytes(&mut buf).ok_or_else(|| unreachable!())?;
+                res.write_hex_buf(&buf)?;
+                res.write_str(",")?;
+                res.write_num(*length)?;
+            }
+            TracepointAction::Expression { expr } => {
+                res.write_str("X ")?;
+                res.write_num(expr.len())?;
+                res.write_str(",")?;
+                res.write_hex_buf(expr)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+/// A list of TracepointActions, either raw and unparsed from a GDB packet, or
+/// a slice of parsed structures like which may be returned from enumerating
+/// tracepoints.
+#[derive(Debug)]
+#[allow(missing_docs)]
+pub enum TracepointActionList<'a, U> {
```

Comment:
The problem is that the API surface we are provided and respond with is `DefineTracepoint`s, not `TracepointActionList`s, and so splitting this enum isn't two separate types entirely instead of an enum is slightly annoying. I would need to parameterize the types like `DefineTracepoint<RawActionList>`/`DefineTracepoint<ParsedActionList>` along with `TracepointItem<RawActionList>`/`TracepointItem<ParsedActionList>`, but felt that was kind of gross. If that's desired it's a change I can do however.